docs/source/developer_guide/generator-runner.rst

```diff
@@ -0,0 +1,358 @@
+If the "Stdin Writer Thread" reads ``None`` from the "Stdin Queue", it will close the stdin-descriptor of the subprocess (usually the writing-end of a pipe-like structure).
+The execution of these threads is independent from the execution of the calling thread, i.e. the thread that executes the ``run``-method of the runner.
+In the figure the calling thread is identified as "Main Thread".
+
+
+If the protocol that is provided to the ``run``-method inherits from ``GeneratorMixIn``, the ``run``-method behaves differently from the synchronous case described in a previous section.
+Instead of blocking, it immediately returns a generator object to the caller, referred to as "Result Generator" in the figure.
+Whenever the caller invokes ``send()`` on the generator, e.g. by iterating over it or by calling ``next(generator)``, the generator will do the following:
+
+1. Check whether its internal result queue, referred to as "Result Queue" in the figure, contains entries.
+   If that is the case, it removes the first entry from the result queue and yields it.
+
+2. As long as the "Result Queue" is empty and the subprocess is still running, it will trigger a single step in the "Queue Reading & Timeout Creation"-logic (implementation remark: this logic is also used in the synchronous execution mode).
+   The "Queue Reading & Timeout Creation"-logic might either block or invoke one of the protocol callbacks.
+   In the figure, the callbacks ``pipe_data_received`` and ``timeout`` are explicitly named, but any callback would be called from here.
+   If the internal result queue is empty, and the process is not running any more, the generator will continue with step 5.
+
+3. The protocol-object might invoke the method ``send_result``, which is defined in the ``GeneratorMixIn``-class.
+   Any object that is passed to ``send_result`` is enqueued into the "Result Queue" of the Generator.
+   To be able to do this, the ``GeneratorMixIn``-objects posses an attribute that points to the internal result queue of the generator.
+
+4. If the internal result queue of the generator is not empty, the generator will yield the first element of the result queue and go back to step 2.
+
+5. The subprocess has exited, the next call to the its ``send()``-method will retrieve the exit status of the subprocess and end the iteration by raising ``StopIteration``.
+
+From the description above and from the figure it should be clear, that, if a user wants to receive output from a subprocess and not just wait for its exit, the user has to send
+data that is received via ``pipe_data_received`` to the result queue of the result generator.
+That means basically he has to call ``send_result()``.
+This is not automatically mixed into the protocol class by ``GeneratorMixIn``.
+A minimal generator protocol would therefore look like this.
+
```

Comment:
I'm not 100% sure, is this something that already exists in the code? Or is it something one would need to implement if this functionality is desired?

Reply:
Yes, that is confusing. The outlined action, i.e. calling `send_result()` in the `pipe_data_received``-callback, is implemented in all provided generator-protocols. But if you create your own protocol implementation, you will have to do it as well.

I will try to make this more clear.

Reply:
Addressed in cd13ec7c13abeb5f10cc584ab7f48c7f13f1785c

Reply:
Done in cd13ec7c13abeb5f10cc584ab7f48c7f13f1785c